robot_smach_states/src/robot_smach_states/startup.py

```diff
@@ -1,148 +1,201 @@
+        return sum(lst) / max(len(lst), 1)
 
     def process_scan(self, scan_msg):
+        """
+        callback function checking the distance of the laser points in front of the robot.
+        :param scan_msg: sensor_msgs.msg.LaserScan
+        :return: no return
+        """
         try:
-            middle_index = len(scan_msg.ranges)/2  # Get the middle point
-            ranges_at_center = scan_msg.ranges[middle_index-2:middle_index+2]  # Get some points around the middle
-            distance_to_door = self.avg(ranges_at_center)  # and the average of the middle range and use it as the distance to the door
+            number_beams = len(scan_msg.ranges)
+            front_index = int(
+                math.floor((self.laser_upside_down*self.laser_yaw - scan_msg.angle_min) / max(scan_msg.angle_increment, 1e-10)))
+
+            if front_index < 2 or front_index > number_beams - 2:
+                rospy.logerr("Base laser can't see in front of the robot")
+                self.laser_sub.unregister()
+                raise IndexError()  # TODO Matthijs: very ugly
+
+            ranges_at_center = scan_msg.ranges[front_index - 2:front_index + 2]  # Get some points around the middle
+            distance_to_door = self.avg(ranges_at_center)  # and the average of the middle range
             rospy.logdebug("AVG distance: {}".format(distance_to_door))
-            self.distances += [distance_to_door] #store all distances
+            self.distances += [distance_to_door]  # store all distances
 
-            avg_distance_now = self.avg(self.distances[-5:]) #And the latest 5
+            avg_distance_now = self.avg(self.distances[-5:])  # And the latest 5
 
             # print "d_start = {0}, d_now = {1}, curr = {2}".format(avg_distance_at_start, avg_distance_now, distance_to_door)
             if avg_distance_now > 1.0:
                 rospy.loginfo("Distance to door is more than a meter")
-                self.door_open.set() #Then set a threading Event that run is waiting for.
+                self.door_open.set()  # Then set a threading Event that execute is waiting for.
         except Exception, e:
             rospy.logerr("Receiving laser failed so unsubscribing: {0}".format(e))
             self.laser_sub.unregister()
 
-    def run(self, robot, timeout):
+    def execute(self, userdata=None):
         rospy.loginfo("Waiting for door...")
-        self.laser_sub = rospy.Subscriber("/"+robot.robot_name+"/base_laser/scan", LaserScan, self.process_scan)
-
-        opened_before_timout = self.door_open.wait(timeout)
+        r = Quaternion()
+        _, (r.x, r.y, r.z, r.w) = self._robot.tf_listener.lookupTransform(self._robot.base_link_frame,
+                                                                          self._robot.robot_name + '/base_laser')
+        laser_rotation = quaternionMsgToKdlRotation(r)
+        self.laser_upside_down = -math.copysign(1, math.cos(laser_rotation.GetRPY()[0]))  # -1 normal, 1 upside down
```

Comment:
What is this needed for?

Reply:
In case a laser is ever not going to be position straight forward

Reply:
Aha. I didn't spot the use of the variable but that's in the callback ```process_scan```